FEATURE: Upgrade demo.neo.io to Neos 9

markusguenther opened the pull request.

markusguenther committed:
TASK: Use localbeach without devkit

markusguenther committed:
TASK: Upgrade dependencies to Neos 9

markusguenther committed:
TASK: Disable new Media-UI

As it is not Neos 9.0 compatible yet

markusguenther committed:
TASK: Adjust UserRegistration to Neos 9 API

markusguenther committed:
TASK: Temporary removal of asset policies

markusguenther commented on DistributionPackages/Neos.Demo.UserRegistration/Configuration/Policy.yaml:
```diff
@@ -1,19 +1,4 @@
 privilegeTargets:
-  'Neos\Media\Security\Authorization\Privilege\ReadAssetPrivilege':
-    'Neos.Demo:SeeOwnAssetCollectionContent':
-      matcher: 'isInCollection("context.userInformation.personalWorkspaceName")'
```
Here we need to find another way as the userInformation.personalWorkspaceName does not exists anymore. We need to use it like in the AddUserWorkpaceAction.

dlubitz reviewed (approved):
Looks good to me! Thank you

markusguenther committed:
TASK: Readd Asset and AssetCollection policies with custom ConditionGenerators

markusguenther commented:
@dlubitz I have adjusted the AssetCollection and Asset Policies. Maybe that is worth a review. The reason for that was that we removed ```context.userInformation``` and therfore the policies did not work anymore.

dlubitz commented on DistributionPackages/Neos.Demo.UserRegistration/Classes/Security/Authorization/Privilege/Doctrine/AbstractConditionGenerator.php:
```diff
@@ -0,0 +1,57 @@
+            $contentRepositoryId = ContentRepositoryId::fromString('default');
+            $userWorkspaceName = $this->metadataAndRoleRepository->findWorkspaceNameByUser($contentRepositoryId, $existingUser->getId());
+
+            return !$userWorkspaceName ? '' : $userWorkspaceName->__toString();
```
Shouldn't this be:
```suggestion
            return !$userWorkspaceName ? '' : $userWorkspaceName->value;
```

dlubitz reviewed (approved):
TBH I don't understand that much regarding the asset condition. But looks good in general. I left one small nitpick.

markusguenther committed:
TASK: Adjust README to use PHP 8.3 container

markusguenther committed:
Update DistributionPackages/Neos.Demo.UserRegistration/Classes/Security/Authorization/Privilege/Doctrine/AbstractConditionGenerator.php

Co-authored-by: Denny Lubitz <lubitz@vivomedia.de>

markusguenther commented:
> TBH I don't understand that much regarding the asset condition. But looks good in general. I left one small nitpick.

No problem, we create a backend user from the frontend Form and the new users need a personal workspace and a asset collection. To make it easy they are names like the user and to be able to see just your asset collection we needed the policies.

Locally it seems to work. Hope that I spotted all issues.

Sebobo reviewed (approved):
Looks good by reading, but I would like to test it locally again.

kdambekalns reviewed (approved):
👍 by 👀

I'd be fine with merging, as soon as @Sebobo did some testing.

mhsdesign reviewed (commented):
Thanks for the upgrade!!

Though i have one request for changes.

I think the whole introduction of new privileges is a lot of code and will be hard to maintain - especially without tests and stuff :D

now as replacement i thought we could simply introduce a custom `context.neosDemo.defaultUserWorkspaceName` helper - basically what the old stuff did just different?

```php
             $accountIdentifier = $account->getAccountIdentifier();
             $existingUser = $this->userService->getUser($accountIdentifier, 'Neos.Neos:Backend');
             $contentRepositoryId = ContentRepositoryId::fromString('default');
             $userWorkspaceName = $this->metadataAndRoleRepository->findWorkspaceNameByUser($contentRepositoryId, $existingUser->getId());
```

or maybe that does not work?

markusguenther committed:
FEATURE: Move workspace code to a dedicated Service

Additionally, switch the base workspace on login to the review workspace.
So it is much more convenient for the test user.
